### DISCUSSION TITLE
settings change doesn’t reflect without reload the window, what can be done in vscode extension for that?

### DISCUSSION THREAD

**Discussion by xkeshav** [microsoft/vscode-discussions#570] [Extension Development QnA] [answered]
created extension, added custom settings but the effect of that setting doesnt reflct immedietely , need to reload the window.

option is to create sparate file , is user check it then a new file will be created but if it not selected ( default ) then add on same file.

I have changed this option before executing the command.

but it is not affecting.
so my question is how do we reload the window or tell your extension to check before execution whether setting change event occured or not?

**Comment by mbehr1** [comment]
see onDidChangeConfiguration: Event <https://code.visualstudio.com/api/references/vscode-api#Event><ConfigurationChangeEvent <https://code.visualstudio.com/api/references/vscode-api#ConfigurationChangeEvent>>

from e.g. https://code.visualstudio.com/api/references/vscode-api and see from the event whether it affects your setting  ev.affectsConfiguration(…)

**Reply by xkeshav** [reply]
I am able to catch the config change event but how do I inform my extension about the change so that on execution of command, it adhere to the change?

**Comment by alefragnani** [comment]
Hi @xkeshav ,

You can be notified about settings changes, if you handle `vscode.workspace.onDidChangeConfiguration`. 

Whenever a setting is updated, the this event is fired and you can upload your extension accordingly.

```
context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(cfg => {
        if (cfg.affectsConfiguration("workbench.iconTheme")) {
            providerManager.refreshTreeViews()
        }

        if (cfg.affectsConfiguration("projectManager.sortList")) {
            updateSortByContext();
            providerManager.storageProvider.refresh();
            providerManager.updateTreeViewStorage();
        }
    }));
```

You can hear any setting, even from other extensions btw.

Hope this helps

**Reply by xkeshav** [reply]
what is provideManager here?

**Reply by alefragnani** [reply]
Sorry, this is from one of my extensions. It's just a snippet for you to see how the event itself works. 

You can ignore it, and think this as:

```ts
context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(cfg => {
        if (cfg.affectsConfiguration("someSetting")) {
            // do your things
        }

        if (cfg.affectsConfiguration("someOtherSetting")) {
            // do other things
        }
    }));
```

**Reply by xkeshav** [reply]
I did this but it is not returning the latest value but provide previous value
```js
	context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(cfg => {
		const affected = cfg.affectsConfiguration('cssColorCollector.colorInSeparateFile');
		if (affected) {
			const newConfig = config.get('cssColorCollector.colorInSeparateFile');
			console.log({newConfig});
		}
	}));
```

as we see the setting checked and visible in workspace setting view but how does it not updated here and do I really need to reload the window which is bad expereince or there is mechanism to get the latest one.

**Reply by xkeshav** [reply]
and I also notice it invokd whenever config option changed, whether in **User settings** or **Workspace settings** then how do we identify the source of it or how do we read only workspace setting/user settings

**Reply by alefragnani** [reply]
If you need to know the setting _from some specific level_ (User, Workspace, Folder) you must use `WorkspaceConfiguration.inspect` instead, as the `getConfiguration.get` API _merges_ the values.

**Reply by alefragnani** [reply]
About the `onDidChangeConfiguration` not reflecting the _new value_, I don't remember issues on the API myself. It was always returning the new value.

Are you updating the setting via Settings UI or JSON?

**Reply by xkeshav** [reply]
via Settings UI. 

figured it out.  

I have assigned getConfiguration() in `config` variable 
```js
const config = vscode.workspace.getConfiguration();
```
and then hecking with `config` variable ,

```js
 config.get('cssColorCollector.colorInSeparateFile');
```
which I guess holds the previous config value when invoked.

so now checking directly 

`vscode.workspace.getConfiguration().get('setting.name');`

then it gives the new updated config value

```js
	context.subscriptions.push(vscode.workspace.onDidChangeConfiguration(cfg => {
		const affected = cfg.affectsConfiguration('cssColorCollector.colorInSeparateFile');
		if (affected) {
			const newConfig = vscode.workspace.getConfiguration().get('cssColorCollector.colorInSeparateFile'); // here
			console.log({newConfig});
		}
		}));

```


Thanks for the help. 

btw I really admire your extension "project manager" , even mentioned in my [extension README ](https://github.com/xkeshav/color-collector#readme)too :)

**Reply by alefragnani** [reply, marked as answer]
Great to see it is working now, and thank you for your kind words 🤝
